src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -788,6 +792,38 @@ public static double IEEERemainder(double x, double y)
             }
         }
 
+        public static int ILogB(double x)
```

Comment:
There is some additional handling in `src/coreclr/jit/valuenum.cpp` that supports constant folding. We need to ensure that stays in sync with the managed implementation.

I'm unsure if there is a trivial way to get value numbering to defer to the managed implementation or if instead we need to maintain two copies of the code. @jkotas or @briansull may be able to help on this part.

Reply:
It should be possible to defer to the managed implementation (for both JIT and AOT cases). It will require a new method on JIT/EE interface.

Reply:
Just as a summary...

With this change, we'll have 3 `System.Math/MathF` methods implemented in managed code:
* `Round`
* `ScaleB`
* `ILogB`

The first has a corresponding native implementation with CSE support, while the second only has a managed implementation and no CSE support.

One of the main concerns with moving all methods to managed is long term maintenance cost, especially around more complex methods like `sin` or `cos`. Hence we've done minimal investigation around having a single native implementation for all platforms taken from something like https://github.com/amd/aocl-libm-ose.

However, we do have some methods (like `ILogB`, `ScaleB`, and `Round`) that are "sufficiently simple" that implementing in managed code is probably worthwhile. It would be nice to ensure that we consistently handle them in CSE when we do move them to managed so we don't lose out on any existing optimizations and so we don't need to maintain 2 or more copies of the logic.

Reply:
Note that for CSE logic specifically you do not need to do anything special w.r.t. to where and how the method is implemented - you can have it be an opague `VNFunc` and it'll work just fine (this is how a few cast helpers are implemented today).

The only place where the actual implementation matters is VN folding.

Reply:
> Round, ScaleB, ILogB

Are typical implementations of these methods 100% precise, or are they approximations? I am thinking we really only need to solve this issue once we get to the managed implementations of methods that are implemented as approximation (sin, cos, etc.)

Reply:
`Round` is 100% precise.

I'm unsure of the guarantees of `ScaleB` or `ILogB` that we have, but I would expect these two in particular to be 100% precise as well.

Reply:
The managed `ScaleB` and `ILogB` implementations are based off the musl implementations, which are 100% precise to my knowledge.

I was unaware that CSE might treat these functions differently, please let me know if there are any other changes I can make to help keep the implementations consistent.

Reply:
A more precise statement would be that we [fold](https://github.com/dotnet/runtime/blob/0c200955af77bf17cd05488db96d3cdb3716a16c/src/coreclr/jit/valuenum.cpp#L5014) them in VN and thus need the native `ilogb` to always match the managed version. If both are precise (i. e. always return the same value for the same input), then we need not worry about it.

Reply:
@SingleAccretion Thanks, that makes more sense. Since the musl implementations are 100% precise, I don't expect any mismatches. 

Please let me know if there is anything else I can look at.